quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -160,671 +164,714 @@ fn mul_matches(left_ax: &Expression, right_ax: &Expression) -> bool {
+            (
+                ref other,
+                InfixOperator::Plus,
+                Expression::Prefix(PrefixExpression {
+                    operator: PrefixOperator::Minus,
+                    ref expression,
+                }),
+            )
+            | (
+                Expression::Prefix(PrefixExpression {
+                    operator: PrefixOperator::Minus,
+                    ref expression,
+                }),
+                InfixOperator::Plus,
+                ref other,
+            ) => simplify_infix(other, InfixOperator::Minus, expression, limit - 1),
+
+            // Subtraction with negation
+
+            // a - (-b) = a + b
+            (
+                ref left,
+                InfixOperator::Minus,
+                Expression::Prefix(PrefixExpression {
+                    operator: PrefixOperator::Minus,
+                    ref expression,
+                }),
+            ) => simplify_infix(left, InfixOperator::Plus, expression, limit - 1),
+
+            // -expression - right = smaller of [(-expression) - right, -(expression + right)]
+            (
+                ref left @ Expression::Prefix(PrefixExpression {
+                    operator: PrefixOperator::Minus,
+                    ref expression,
+                }),
+                InfixOperator::Minus,
+                ref right,
+            ) => {
+                let original = sub!(left.clone(), right.clone());
+                let new = simplify_prefix(
+                    PrefixOperator::Minus,
+                    &simplify_infix(expression, InfixOperator::Plus, right, limit - 1),
+                    limit - 1,
+                );
+                min_by_key(original, new, size)
+            }
+
+            // Multiplication with negation
+
+            // Double negative: (-a) * (-b) = a * b
+            (
+                Expression::Prefix(PrefixExpression {
+                    operator: PrefixOperator::Minus,
+                    expression: ref a,
+                }),
                 InfixOperator::Star,
-                expression,
-                limit.saturating_sub(1),
-            );
-            min_by_key(original, new, size)
-        }
-        (
-            ref left @ Expression::Prefix(PrefixExpression {
-                operator: PrefixOperator::Minus,
-                ref expression,
-            }),
-            InfixOperator::Star,
-            ref right,
-        ) => {
-            let original = mul!(left.clone(), right.clone());
-            let neg_right = simplify_prefix(PrefixOperator::Minus, right, limit.saturating_sub(1));
-            let new = simplify_infix(
-                expression,
+                Expression::Prefix(PrefixExpression {
+                    operator: PrefixOperator::Minus,
+                    expression: ref b,
+                }),
+            ) => simplify_infix(a, InfixOperator::Star, b, limit - 1),
+
+            // a * (-b) = (-a) * b, pick the shorter
```

Comment:
Could be merged with the case after

Reply:
Not quite, I think, as we reconstruct the original expression to compare for size, right?

Reply:
Is the size actually different if `left` and `right` are swapped?

Reply:
I can think of a family of examples: suppose `left` is π or some number, while `right` is just a variable. Then pushing the negative on `left` instead will simplify further.